File: Ruby/Endpoint Examples/JSON Payload/translated-pdf-text.rb
#!
# Translate PDF text via two-step JSON flow.

require "json"
require "faraday"
require "faraday/retry"
require "dotenv"

Dotenv.load

API_KEY = ENV["PDFREST_API_KEY"]
abort("Missing PDFREST_API_KEY in .env") if API_KEY.nil? || API_KEY.strip.empty?
API_BASE = (ENV["PDFREST_URL"] || ENV["PDFREST_API"] || "https://api.pdfrest.com").sub(%r{/+$}, "")

pdf_path = ARGV[0]
abort("Usage: ruby translated-pdf-text.rb /path/to/file.pdf") unless pdf_path && File.file?(pdf_path)

filename = File.basename(pdf_path)
file_bytes = File.binread(pdf_path)

begin
  upload_conn = Faraday.new(url: API_BASE) do |f|
    f.request :retry, max: 2, interval: 0.2
    f.adapter Faraday.default_adapter
  end
  upload_resp = upload_conn.post("/upload") do |req|
    req.headers["api-key"] = API_KEY
    req.headers["content-filename"] = filename
    req.headers["Content-Type"] = "application/octet-stream"
    req.body = file_bytes
  end
  STDERR.puts upload_resp.body
  abort("Upload failed with status #{upload_resp.status}") unless upload_resp.success?
  upload_json = JSON.parse(upload_resp.body)
  uploaded_id = upload_json.fetch("files").fetch(0).fetch("id")
  STDERR.puts "Successfully uploaded with an id of: #{uploaded_id}"

  conn = Faraday.new(url: API_BASE) do |f|
    f.request :retry, max: 2, interval: 0.2
    f.adapter Faraday.default_adapter
  end
  # Translates text to American English. Format the output_language as a 2-3 character ISO 639 code, optionally with a region/script (e.g., 'en', 'es', 'zh-Hant', 'eng-US').
  body = { id: uploaded_id, output_language: "en-US" }.to_json
  resp = conn.post("/translated-pdf-text") do |req|
    req.headers["api-key"] = API_KEY
    req.headers["Content-Type"] = "application/json"
    req.body = body
  end
  puts resp.body
  abort("Translate failed with status #{resp.status}") unless resp.success?
rescue KeyError => e
  abort("Unexpected response format: #{e.message}")
rescue => e
  abort("Error: #{e.class}: #{e.message}")
end

File: VB.NET/Endpoint Examples/JSON Payload/translated-pdf-text.vb
'''
' Translate PDF text using pdfRest.
' Two-step JSON flow: upload then call /translated-pdf-text with the uploaded id.
'''

Option Strict On
Option Explicit On

Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Net.Http
Imports System.Net.Http.Headers
Imports System.Text
Imports System.Text.Json
Imports System.Threading.Tasks

Namespace VBNetSamples.Endpoint_Examples.JSON_Payload
    Module TranslatedPdfText
        Public Async Function Execute(args As String()) As Task
            If args Is Nothing OrElse args.Length < 1 Then
                Console.Error.WriteLine("Usage: dotnet run -- translated-pdf-text /path/to/input.pdf")
                Environment.Exit(1)
            End If

            Dim inputPath As String = args(0)
            If Not File.Exists(inputPath) Then
                Console.Error.WriteLine($"Input file not found: {inputPath}")
                Environment.Exit(1)
            End If

            Dim apiKey As String = Environment.GetEnvironmentVariable("PDFREST_API_KEY")
            If String.IsNullOrWhiteSpace(apiKey) Then
                Console.Error.WriteLine("Missing environment variable PDFREST_API_KEY.")
                Environment.Exit(1)
            End If

            Dim baseUrl As String = Environment.GetEnvironmentVariable("PDFREST_URL")
            If String.IsNullOrWhiteSpace(baseUrl) Then baseUrl = "https://api.pdfrest.com"

            Dim baseUri As Uri
            Try
                baseUri = New Uri(baseUrl)
            Catch ex As Exception
                Console.Error.WriteLine($"Invalid PDFREST_URL: {baseUrl}")
                Environment.Exit(1)
                Return
            End Try

            Using httpClient As New HttpClient()
                httpClient.BaseAddress = baseUri

                ' 1) Upload
                Dim uploadRequest As New HttpRequestMessage(HttpMethod.Post, "upload")
                uploadRequest.Headers.TryAddWithoutValidation("Api-Key", apiKey)
                uploadRequest.Headers.Accept.Add(New MediaTypeWithQualityHeaderValue("application/json"))

                Dim fileBytes As Byte() = File.ReadAllBytes(inputPath)
                Dim uploadContent As New ByteArrayContent(fileBytes)
                uploadContent.Headers.TryAddWithoutValidation("Content-Type", "application/octet-stream")
                uploadContent.Headers.TryAddWithoutValidation("Content-Filename", Path.GetFileName(inputPath))
                uploadRequest.Content = uploadContent

                Dim uploadResponse As HttpResponseMessage = Await httpClient.SendAsync(uploadRequest)
                Dim uploadBody As String = Await uploadResponse.Content.ReadAsStringAsync()
                If Not uploadResponse.IsSuccessStatusCode Then
                    Console.Error.WriteLine($"Upload failed: {CInt(uploadResponse.StatusCode)} {uploadResponse.ReasonPhrase}")
                    Console.Error.WriteLine(uploadBody)
                    Environment.Exit(1)
                End If
                Console.WriteLine(uploadBody)

                Dim uploadedId As String = Nothing
                Try
                    Using doc As JsonDocument = JsonDocument.Parse(uploadBody)
                        uploadedId = doc.RootElement.GetProperty("files")(0).GetProperty("id").GetString()
                    End Using
                Catch ex As Exception
                    Console.Error.WriteLine("Failed to parse upload response JSON for file id.")
                    Console.Error.WriteLine(ex.Message)
                    Environment.Exit(1)
                End Try

                ' 2) Translate via JSON payload
                Dim req As New HttpRequestMessage(HttpMethod.Post, "translated-pdf-text")
                req.Headers.TryAddWithoutValidation("Api-Key", apiKey)
                req.Headers.Accept.Add(New MediaTypeWithQualityHeaderValue("application/json"))

                Dim payload As New Dictionary(Of String, Object) From {
                    {"id", uploadedId},
                    {"output_language", "en-US"} ' Translates text to American English. Format the output_language as a 2-3 character ISO 639 code, optionally with a region/script (e.g., 'en', 'es', 'zh-Hant', 'eng-US').
                }
                Dim payloadJson As String = JsonSerializer.Serialize(payload)
                req.Content = New StringContent(payloadJson, Encoding.UTF8, "application/json")

                Dim resp As HttpResponseMessage = Await httpClient.SendAsync(req)
                Dim body As String = Await resp.Content.ReadAsStringAsync()
                If Not resp.IsSuccessStatusCode Then
                    Console.Error.WriteLine($"Translate request failed: {CInt(resp.StatusCode)} {resp.ReasonPhrase}")
                    Console.Error.WriteLine(body)
                    Environment.Exit(1)
                End If
                Console.WriteLine(body)
            End Using
        End Function
    End Module
End Namespace

